peps/pep-9999.rst

```diff
@@ -157,36 +157,406 @@ splitting off headers and other build-time dependencies in a separate
-- A plugin design, where system package managers can register their name mapping
-- Name mapping should be in a structured, human-readable file format (TBD:
-  JSON, YAML, or ...)
+The mapping infrastructure should include the following components and properties:
+
+- A central registry for known generic and virtual PURLs.
+- A list of known ecosystems and package managers, where ecosystem maintainers
+  can register their name mapping(s).
+- A standardized schemsa that defines how mappings should be structured.
+- The above documents should be written in a structured, human-readable file format.
+  As long as it can be validated with the corresponding JSON Schemas, they could be
+  delivered as JSON, YAML, TOML and/or others. Authors recommend JSON or YAML.
+- One central Python package for the central registry, list of ecosystems and known
+  mappings.
+
+On the client side, there should be:
+
 - A way for the system to specify a default (e.g., the Python install on Ubuntu
   could register ``apt`` as the default system package manager with the
-  registry tool. It may also be left unspecified.
-- A way for the user to specify the default and/or current system package
-  manager. E.g., a user on Ubuntu may want either ``apt``, ``conda``, ``brew``
-  or ``spack`` as their package manager of choice to provide external
-  dependencies.
+  registry tool). It may also be left unspecified.
+- A way for the user to specify the default and/or current system package manager.
+  E.g., a user on Ubuntu may want either ``apt``, ``conda``, ``brew``  or ``spack``
+  as their package manager of choice to provide external dependencies.
 
 
 Specification
 =============
 
-
+Three schemas are proposed:
+
+1. A central registry of PURLs.
+2. A list of known ecosystems and their package manager names.
+3. The ecosystem-specific mappings of PURLs to package specifiers,
+   plus details of their package managers.
+
+The central registry defines which identifiers are recognized as canonical. Each entry MUST
+provide a valid PURL in the field ``id``, with an optional free form ``description`` text.
+Additionally some entries MAY refer to another entry via the ``provides`` field, which takes a
+list of strings already defined as ``id`` in the registry. This is useful for aliases
+(e.g. ``dep:generic/arrow`` and ``dep:github/apache/arrow``), and concrete implementations of
+a ``dep:virtual/`` entry (e.g. ``dep:generic/gcc`` would provide ``dep:virtual/compiler/c``).
+Entries without ``provides`` content or, if populated, only with ``dep:virtual/`` identifiers,
+are considered canonical.
+
+The list of known ecosystems assigns an identifier to each ecosystem and reports the
+canonical location for its mapping. The known package managers are also provided here
+so tools can map package manager names back to ecosystems without querying each mapping.
+
+.. JRG: Not sure if tying the package manager list to the known ecosystems list is worth it.
```

Comment:
Seems necessary to me. If a user says "I want to use micromamba", how would say `uv` know what to do with that information if it couldn't map the package manager name to the conda packaging ecosystem?

Reply:
I mean it in the sense of "I can always take the ecosystem name, fetch its mapping, and find out which package managers are available". Right now we are storing the package manager list together with the known ecosystems list, which means we are centralizing this information. If a given ecosystem adds a new package manager on their mapping, then the central registry would be outdated. Folks can always fetch the mapping to make sure, but that's what I'm wondering: is it worth it to "cache" the package manager list in the known ecosystem list?

Reply:
Ah okay. Perhaps not that important. I don't have a strong intuition. Can you make the choice here?

Reply:
I think that a couple extra requests won't be that important and that if tools are concerned about it, they can cache locally. If we don't do it like that, and maintenance of the centralized mapping stalls or something, then the lists might get out of sync and cause more trouble than it's worth.

Reply:
Opened https://github.com/rgommers/external-deps-build/pull/4

Reply:
thanks!